builder/dockerfile/copy.go

```diff
@@ -405,26 +407,35 @@ func performCopyForInfo(dest copyInfo, source copyInfo, options copyFileOptions)
 		// is a symlink
 		destPath = filepath.Join(destPath, filepath.Base(source.path))
 	}
-	return copyFile(archiver, srcPath, destPath)
+	return copyFile(archiver, srcPath, destPath, options.chownPair)
 }
 
-func copyDirectory(archiver *archive.Archiver, source, dest string) error {
+func copyDirectory(archiver *archive.Archiver, source, dest string, chownPair idtools.IDPair) error {
+	destExists, err := isExistingDirectory(dest)
+	if err != nil {
+		return errors.Wrapf(err, "failed to query destination path")
+	}
 	if err := archiver.CopyWithTar(source, dest); err != nil {
 		return errors.Wrapf(err, "failed to copy directory")
 	}
-	return fixPermissions(source, dest, archiver.IDMappings.RootPair())
+	if !destExists {
+		// if the destination didn't exist, we need to special case chown it
+		// to the requested --chown flags as provided to the original command
+		if err := os.Lchown(dest, chownPair.UID, chownPair.GID); err != nil {
+			return errors.Wrapf(err, "failed to chown target directory")
+		}
```

Comment:
except that it doesn't work for the chown case. I took the longest time trying to decide it if was worth adding complexity to `fixPermissions()` or not. The case that doesn't work is when the new directory is actually created by `archive.CopyWithTar`, so when checked for existence in `fixPermissions`, it *does* exist, so the wrong codepath is taken for the chown case. What I wanted was to be able to make an archiver with `ChownOpts` just for this `ADD` or `COPY` command, but after spending time in that code, it isn't possible without significant changes to add `TarOptions` cascading through a bunch of functions that don't take it today.